### ISSUE TITLE
tf.gradients returns incorrect results if used multiple times on CudnnLSTM

### ISSUE THREAD

**Issue by chrisc36** [tensorflow/tensorflow#26331]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below):1.12.0
- Python version: 3.6.7
- Bazel version (if compiling from source): unknown
- GCC/Compiler version (if compiling from source): unknown
- CUDA/cuDNN version:  10.0 / 7.4
- GPU model and memory: TI 1080, 12GB RAM

I found some cases where fetching multiple gradient computations for the parameters of a CudnnLSTM seems to result in incorrect results. For example, the following code:

```
import tensorflow as tf
from tensorflow.contrib.cudnn_rnn.python.layers.cudnn_rnn import CudnnLSTM

x = tf.fill((2, 1, 5), 0.5)
layer = CudnnLSTM(1, 5)
layer.build(x.shape)
output = layer.apply(x)[0]

sess = tf.Session()

loss = tf.reduce_sum(output)
vars = tf.trainable_variables()
grad1 = tf.gradients(loss, vars)[0]
grad2 = tf.gradients(loss, vars)[0]

sess.run(tf.global_variables_initializer())
g1, g2 = sess.run([grad1, grad2])
print(g1[:10])
print(g2[:10])
print((g1 - g2)[:10])
```
Shows the two gradient computation are far from equal, even though they are requesting exactly the same value. On my machine I saw:

```
[ 0.0290555   0.0290555   0.0290555   0.0290555   0.0290555  -0.00758932
 -0.00758932 -0.00758932 -0.00758932 -0.00758932]
[1.7652620e-04 1.7652620e-04 1.7652620e-04 1.7652620e-04 1.7652620e-04
 7.4650106e-06 7.4650106e-06 7.4650106e-06 7.4650106e-06 7.4650106e-06]
[ 0.02887898  0.02887898  0.02887898  0.02887898  0.02887898 -0.00759678
 -0.00759678 -0.00759678 -0.00759678 -0.00759678]
```

which seems far too large for the difference to be a rounding issue.

Using `g1, g2 = [sess.run(grad1), sess.run(grad2)]` will fix the issue.

The issue also goes away if we use  `grads = sess.run([grad1, grad1])`, but seems to remain for other, non-trivial cases. For example, if we had done `grad2 = tf.gradients(loss*2, vars)[0]` `g2` should be twice as large as `g1`, but it will be way off.

**Comment by qlzh727** [comment]
Adding @protoget who previously worked on cudnn kernels.

**Comment by rmothukuru** [comment]
@protoget,
Sorry for the delayed response. As **`CudnnLSTM layer`** has been depcreated in **`Tensorflow 2.x`**, if possible, can you please try with [RNN or LSTM or GRU with CuDNN Kernels](https://www.tensorflow.org/guide/keras/rnn#performance_optimization_and_cudnn_kernels) and let us know if this issue still persists? Thanks!
